### PR TITLE
Extend auto shard capabilities to work around torch.fx edge cases.

### PR DESCRIPTION
**Pull request by EugenHotaj**
auto_shard.py currently uses torch.fx to create a symbolic DAG of
operations and linearizes that DAG into an nn.Sequential so it can later
be used for model offloading. This works in most cases but runs into
issues for certain eager mode features, such as dynamic conditionals,
shape-dependent computation, etc.

This PR extends auto_shard.py to first run a preprocessing step which wraps
any nn.Module which cannot be traced through. It adds a test for dynamic
conditionals and updates existing failing test code.

There are some immediate extensions to this approach which are marked as
TODO in the code.

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/main/CONTRIBUTING.md)?
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [x] N/A
- [x] Did you make sure to update the docs?
  - [x] N/A
- [x] Did you write any new necessary tests?
  - [ ] N/A
- [x] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/main/CHANGELOG.md)? (if needed)
  - [x] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: tests/experimental/nn/test_auto_shard.py [hunk 1]
```diff
@@ -107,3 +105,39 @@ def test_single_run():
...
+        return x
+
+
+def test_dynaimc_conditionals_auto_wrapped():
+    if torch_version() < (1, 8, 0):
+        pytest.skip("requires torch version >= 1.8.0")
+    from fairscale.experimental.nn.auto_shard import shard_model
+
+    model = BranchedNetwork()
+    sharded_model = shard_model(model, 3)
```
Comment (anj-s) [inline tests/experimental/nn/test_auto_shard.py:140]: can we test a quick FW pass to make sure we see the same result when compared to unsharded?

File: tests/experimental/nn/test_auto_shard.py [hunk 2]
```diff
@@ -107,3 +105,39 @@ def test_single_run():
...
+    # TODO(ehotaj): There might be a bug in our split code because we shard the
+    # model into 10 shards even though we specify 3 shards above.
+    assert len(sharded_model) == 10
```
Comment (anj-s) [inline tests/experimental/nn/test_auto_shard.py:143]: Can you print out the original model and sharded model?

Comment (EugenHotaj) [reply tests/experimental/nn/test_auto_shard.py:143]: Here is the full model:

```
BranchedNetwork(
  (net): ModuleList(
    (0): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
    (1): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
    (2): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
    (3): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
    (4): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
    (5): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
    (6): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
    (7): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
    (8): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
    (9): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
  )
)
```

And the branched model:

```
[GraphModule(
  (net): Module(
    (0): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
  )
), GraphModule(
  (net): Module(
    (1): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
  )
), GraphModule(
  (net): Module(
    (2): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
  )
), GraphModule(
  (net): Module(
    (3): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
  )
), GraphModule(
  (net): Module(
    (4): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
  )
), GraphModule(
  (net): Module(
    (5): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
  )
), GraphModule(
  (net): Module(
    (6): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
  )
), GraphModule(
  (net): Module(
    (7): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
  )
), GraphModule(
  (net): Module(
    (8): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
  )
), GraphModule(
  (net): Module(
    (9): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
  )
)]
```

Looks reasonable to me but I'm not sure why the `shard_count` is not being respected.

File: tests/experimental/nn/test_auto_shard.py (hunk 1 above)
Comment (EugenHotaj) [reply tests/experimental/nn/test_auto_shard.py:142]: Done.

File: tests/experimental/nn/test_auto_shard.py (hunk 2 above)
Comment (EugenHotaj) [reply tests/experimental/nn/test_auto_shard.py:145]: > I'm not sure why the shard_count is not being respected.

BTW, this also happens without the changes in this PR, i.e. if we just use `torch.fx.symbolic_trace` directly, so I think something might be going wrong in the `_split_nodes` logic. 

(I could try looking into it in a follow up PR).

**Pull request merged**